### DISCUSSION TITLE
When the /etc/hosts file changes, how does kong dynamically load these changes so that the service can be routed normally?

### DISCUSSION THREAD

**Discussion by shilaidun** [Kong/kong#13105] [Help]
1.start kong 
2.`echo '1.1.1.1 test1' >> /etc/hosts`
3.add service `test1:8080` in kong 
4.curl http://127.0.0.1:8000/test1/xxx.do 
5.response : `{
"message": "failure to get a peer from the ring-balance"
}`
6.kong.error.log : `[warn] 255829#0: *532 [lua] targets.lua:504: queryDns(): querying dns for test1 failed: dns server error: 3 name error. Tried ["(short)test1:(na) - cache-hit/stale","test1:1 - cache-hit/dns server error: 3 name error","test1:33 - cache-miss/querying/dns server error: 3 name error","test1:5 - cache-miss/querying/dns server error: 3 name error"], context: ngx.timer`
7.please help me

**Comment by chronolaw** [comment]
@chobits , do you mind to take a look? thanks.

**Reply by chobits** [reply]
You might need to restart kong to let it load /etc/hosts

**Reply by shilaidun** [reply]
DNS can configure TTL, why can’t `/etc/hosts`

**Reply by chobits** [reply]
I mean, you can configure /etc/hosts, but you need to restart Kong to have it load the changes. Both /etc/hosts and kong.conf serve as static configurations for Kong, and only a Kong restart will apply the changes.

From your steps to reproduce, we didnt see you restart kong after modifying /etc/hosts

**Reply by shilaidun** [reply]
I mean, configure `dns_valid_ttl` or other DNS settings can reload DNS when cache  does not exist in kong, why is `/etc/hosts` not reloaded all together.

**Reply by chobits** [reply]
> I mean, configure dns_valid_ttl or other DNS settings can reload DNS when cache does not exist in kong, why is /etc/hosts not reloaded all together.

I cannot understand what you said. Can you use a tool similar to GPT to translate your native language into English?

You mean why dns_valid_ttl does not affect entries of /etc/hosts?
The dns_valid_ttl configuration parameter in Kong determines how long DNS entries remain valid in Kong's internal cache. However, entries in /etc/hosts are not managed or controlled by DNS; they are static mappings of hostnames to IP addresses that bypass DNS resolution entirely. Therefore, entries from /etc/hosts do not fall under the purview of dns_valid_ttl and are not affected by it.

**Reply by shilaidun** [reply]
Can a configuration parameter be added to kong.conf to dynamically refresh the file specified by the `dns_hostsfile` configuration, such as the default `/etc/hosts`?

**Reply by chobits** [reply]
No, introducing this feature would complicate the Kong DNS client unnecessarily. I believe the DNS protocol already supports this well. You can store 'test1' in the DNS server.
